design-proposals/klip-9-table-functions.md

```diff
@@ -0,0 +1,206 @@
+in messages that they wish to explode. This new feature would allow them to do that in a pure KSQL
+app thus allowing them to create richer KSQL applications.
+
+## Public APIs
+
+There will be no changes to the KSQL language or any existing public APIs or configuration files.
+
+We will introduce new annotations to allow users to mark their UDTFs, in a very similar way to
+UDAFs.
+
+## Design
+
+### General plumbing
+
+We will need the following general plumbing to support table functions:
+
+* In a similar way to how UDAFs are currently supported we will need to create a 
+`KsqlTableFunction` which is a wrapper around the actual table function (whether built-in or
+generated). We'll also need to augment the function registry to manage the addition and retrieval
+of table functions.
+* Augment the query analysis code to extract information relating to any table functions
+in the query. We can do this either by extending the existing query analyzer or perhaps creating
+a new `TableFunctionAnalyzer` in a similar way to how aggregate analysis uses its own 
+`AggregateAnalyzer`.
+* Augment the building of logical plan by building a new node representing a FlatMap stage of the
+stream processing in the case there are table functions in the SELECT.
+* Augment the execution plan logic by adding new nodes in the execution plan corresponding to any
+extra FlatMap stage required by the query.
+* Amend the columns used in the final projection to reflect any generated columns created by a 
+FlatMap stage.
+* Implement the actual FlatMap logic. For each row, any table functions on that row will be evaluated.
+The number of rows returned from the FlatMap will be equal to the greatest of the number of rows
+returned from any of the table functions for that row. Any non table function values for the row will
+be duplicated for each row. For any table function that returned fewer than the greatest number of rows
+the values for that columnn will be `null`. For performance reasons, the FlatMap logic should cache any
+ table functions for the columns to avoid them being looked up each time.
+
+### Implement set of built in table functions
+
+We will implement the following built in table functions:
+
+* `EXPLODE(ARRAY<V>)` -> Results in `N` rows (where `N` is number of elements in array) each with a column
+of type `V` holding the value of the element
+* `EXPLODE(MAP<K, V>)` -> Results in `N` rows (where `N` is number of elements in map) each with a column
+holding a `STRUCT<key K, value V>` that contains the key and value.
+* `INDEX_EXPLODE(ARRAY<V>)` -> Results in `N` rows (where `N` is number of elements in array) each with a
+column of type `STRUCT<index INT, value V>` containing the index of the element from the original array
+and the value from the array.
+
+### Implement framework of user defined table functions
+
+We will introduce annotations `@UdtfFactory` and `@UdtfDescription` in a similar way to UDAFs.
```

Comment:
As per @agavra comment - I don't think we need a factory.  We should go more inline with UDFs.

And example UDTF here would help people to get a grip on what is being proposed.

Reply:
I published a working prototype last well, with an example UDTF and working QTT test:

https://github.com/purplefox/ksql/pull/6